### PR TITLE
Extending FloatingWatermarks

### LINKED ISSUE #2486: DateTimePicker right click menu appearance
## What steps will reproduce this issue?

The appearance of the context menu is not correct

result

### Expected outcome

The context menu should be this way
expected

Te code:
`<Controls:DateTimePicker SelectedDate="{Binding DatePickerDate, Mode=TwoWay}" IsReadOnly="{Binding ReadOnly}" PickerVisibility="All" />`

--
### Environment

- MahApps.Metro v1.3.0-ALPHA137 (Prerelease)
- Windows 7
- Visual Studio 2013
- .NET Framework 4.5

### LINKED ISSUE #2524: DatePicker and DateTimePicker UseFloatingWatermark don't work
## What steps will reproduce this issue?
````xaml
<DatePicker  Controls:TextBoxHelper.UseFloatingWatermark="True"
             Controls:TextBoxHelper.Watermark="Watermark:"  />
````
If SelectedDate is null, then Watermark is visible, but FloatingWatermark don't work when SelectedDate != null.

### Expected outcome

Should work like on others controls

--
### Environment

- MahApps.Metro __v1.3.0.155__
- Windows __10__
- Visual Studio __2013 & 2015__
- .NET Framework __4.5__

### PR DESCRIPTION
**Pull request by xxMUROxx**
- [x] DatePicker
- [x] DateTimePicker
- [x] TimePicker

-
- [x] Final Testing

Resolves  #2524 
Resolves #2486

### CODE REVIEW

File: MahApps.Metro/Controls/Helper/TextBoxHelper.cs [hunk 1]
```diff
@@ -22,7 +22,6 @@ public class TextBoxHelper
         public static readonly DependencyProperty IsMonitoringProperty = DependencyProperty.RegisterAttached("IsMonitoring", typeof(bool), typeof(TextBoxHelper), new UIPropertyMetadata(false, OnIsMonitoringChanged));
         public static readonly DependencyProperty WatermarkProperty = DependencyProperty.RegisterAttached("Watermark", typeof(string), typeof(TextBoxHelper), new UIPropertyMetadata(string.Empty));
         public static readonly DependencyProperty UseFloatingWatermarkProperty = DependencyProperty.RegisterAttached("UseFloatingWatermark", typeof(bool), typeof(TextBoxHelper), new FrameworkPropertyMetadata(false, ButtonCommandOrClearTextChanged));
-        public static readonly DependencyProperty TextLengthProperty = DependencyProperty.RegisterAttached("TextLength", typeof(int), typeof(TextBoxHelper), new UIPropertyMetadata(0));
```
Comment (punker76) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @xxMUROxx why removing this? it's a breaking change...

File: MahApps.Metro/Behaviours/DatePickerTextBoxBehavior.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    using System.Windows.Interactivity;
+    using MahApps.Metro.Controls;
+
+    public class DatePickerTextBoxBehavior : Behavior<DatePickerTextBox>
+    {
+        protected override void OnAttached()
+        {
+            base.OnAttached();
+            AssociatedObject.TextChanged += OnTextChanged;
+        }
```
Comment (punker76) [inline MahApps.Metro/Behaviours/DatePickerTextBoxBehavior.cs]: @xxMUROxx this is a possible mem leak

File: MahApps.Metro/Controls/Helper/TextBoxHelper.cs (hunk 1 above)
Comment (xxMUROxx) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @punker76 Thanks for the review. I thought that this property is not used and (in my case) did not find any reason to use this property.
But if you want, I revert this commit

Comment (punker76) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @xxMUROxx I'm with you, but it's a breaking change if you delete this here in dev branch for v1.3.0 This attached property makes no sense for all our custom input controls (NumericUpDown...)

Comment (xxMUROxx) [inline MahApps.Metro/Controls/Helper/TextBoxHelper.cs]: @punker76 Ok. I revert the Change for 1.3

File: MahApps.Metro/Behaviours/DatePickerTextBoxBehavior.cs (hunk 2 above)
Comment (xxMUROxx) [inline MahApps.Metro/Behaviours/DatePickerTextBoxBehavior.cs]: @punker76 any suggestions how to improve it?

WeakEventManager<DatePickerTextBox, TextChangedEventArgs>.AddHandler(AssociatedObject, TextBoxBase.TextChangedEvent.Name, OnTextChanged);

is only available for NET45

Comment (thoemmi) [inline MahApps.Metro/Behaviours/DatePickerTextBoxBehavior.cs]: You've forgot to override `OnDetaching`, like
```c#
protected override void OnDetaching()
{
    AssociatedObject.TextChanged -= OnTextChanged;
    base.OnDetaching();
}
```

Comment (xxMUROxx) [inline MahApps.Metro/Behaviours/DatePickerTextBoxBehavior.cs]: @thoemmi thanks for pointing that out. An additional question is: `OnDetaching` is only called when attaching/detaching throug C# code not when attaching the behavior directly throug XAML, like it is done here. Do you think this might be a memory leak to or is this no problem?

**Comment by xxMUROxx** [comment]
@punker76 Ok, for me this PR is finished.
